File: raplbaddi/supportrapl/report/payments_service_center/payments_service_center.js
// Copyright (c) 2024, Nishant Bhickta and contributors
// For license information, please see license.txt

frappe.query_reports["Payments Service Center"] = {
	onload: function(report) {
		// arr = frappe.datetime.get_today().split("-")
        // arr.push("0" + (parseInt(arr.pop()) - 1))
        // yesterday = arr.join("-") 

        report.page.add_inner_button(__("Today"), function() {
            frappe.query_report.get_filter('start_date').set_value(frappe.datetime.get_today());
            frappe.query_report.get_filter('end_date').set_value(frappe.datetime.get_today());
        }, "Range");
        // report.page.add_inner_button(__("Yesterday"), function() {
        //     frappe.query_report.get_filter('start_date').set_value(yesterday);
        //     frappe.query_report.get_filter('end_date').set_value(yesterday);
        // }, "Range");
        report.page.add_inner_button(__("Week"), function() {
            frappe.query_report.get_filter('start_date').set_value(frappe.datetime.week_start());
            frappe.query_report.get_filter('end_date').set_value(frappe.datetime.week_end());
        }, "Range");
		report.page.add_inner_button(__("Month"), function() {
			frappe.query_report.get_filter('start_date').set_value(frappe.datetime.month_start());
			frappe.query_report.get_filter('end_date').set_value(frappe.datetime.month_end());
		}, "Range");
		report.page.add_inner_button(__("Quarter"), function() {
			frappe.query_report.get_filter('start_date').set_value(frappe.datetime.quarter_start());
			frappe.query_report.get_filter('end_date').set_value(frappe.datetime.quarter_end());
		}, "Range");
        report.page.add_inner_button(__("Year"), function() {
            frappe.query_report.get_filter('start_date').set_value(frappe.datetime.year_start());
            frappe.query_report.get_filter('end_date').set_value(frappe.datetime.year_end());
        }, "Range");
    },
	"filters": [
		{
			'fieldname':'Service_Centre',
			'label': __('Service Centre'),
			'fieldtype':'Link',
			'options':'Service Centre'
		},
		{
			'fieldname': 'start_date',
			'label': __('Start Date'),
			'fieldtype': 'Date'
		},
		{
			'fieldname': 'end_date',
			'label': __('End Date'),
			'fieldtype': 'Date'
		},
		{
			'fieldname': 'is_paid',
			'label': __('Paid'),
			'fieldtype': 'Check'
		},
		{
			'fieldname':'group_by_sc',
			'label': __('Group By Service Centre'),
			'fieldtype':'Check',
			'default': 1
		},

	]
};

File: raplbaddi/supportrapl/report/payments_service_center/payments_service_center.py
# Copyright (c) 2024, Nishant Bhickta and contributors
# For license information, please see license.txt

import frappe
class PayementReport:
	def __init__(self, filters):
		self.filters = filters

	def get_data(self):						
		query = f"""
		select 
		i.name AS complaint_no,
		SUBSTRING_INDEX(i.custom_creation_date," ",1) as 'Date', 
		j.service_centre_name as 'Service Center',
		j.bank_name as 'Bank',
		j.bank_account_no as 'Account No.',
		j.ifsc_code as 'Bank IFSC',
		j.upi_id as 'UPI Id',
		i.kilometer as 'KMs',
		i.amount as 'Amount'
		from tabIssueRapl as i
		inner Join 
		`tabService Centre` as j on i.service_centre = j.service_centre_name
		where i.service_delivered = 'Yes'
		"""
		
		result = frappe.db.sql(query, as_dict=True)
		
		self.data = result
        
	def get_columns(self):
		columns = [
		{'fieldname': 'complaint_no','label': ('Complaint No.'),'fieldtype': 'Link','options':'IssueRapl','width':120},
		{'fieldname': 'Date','label': ('Date'),'fieldtype': 'Date','width':120},
		{'fieldname': 'Service Center','label': ('Service Center'),'fieldtype': 'Link','options':'Service Centre','width':180},
		{'fieldname': 'Bank','label': ('Bank Name'),'fieldtype': 'Data','width':150},
		{'fieldname': 'Account No.','label': ('Account No.'),'fieldtype': 'Data','width':130},
		{'fieldname': 'Bank IFSC','label': ('IFSC Code'),'fieldtype': 'Data','width':130},
		{'fieldname': 'UPI Id','label': ('UPI Id'),'fieldtype': 'Data','width':120},
		{'fieldname': 'KMs','label': ('KMs'),'fieldtype': 'Data','width':70},
		{'fieldname': 'Amount','label': ('Amount'),'fieldtype': 'Data','width':90}
		]
		return columns
    
    	
	def get_filtered_data(self):
		start_date = self.filters.get('start_date')
		end_date = self.filters.get('end_date')
		service_centre = self.filters.get('Service_Centre')
		self.get_data()
		self.filtered_data = []
		for i in self.data:
			if (not service_centre or i['Service Center'] == service_centre)and(not start_date or i['Date'] >= start_date) and (not end_date or i['Date'] <= end_date):
				self.filtered_data.append(i)
		return self.filtered_data

	def get_count(self):
		return len(self.filtered_data)


	def get_msg(self):
		count = self.get_count()
		# amount = self.get_total_amount()
		# sales_person = self.filters.sales_person
		return f"""
		<div style="display: flex; justify-content: space-between;">
			<h3>Complaints : {count}</h3>
			
		</div>"""


def execute(filters=None):
	payment = PayementReport(filters)
	columns, data = [], []
	return payment.get_columns(), payment.get_filtered_data(),payment.get_msg()
